Simplify `cupy.linalg.inv`

anaruse opened the pull request:
This PR uses `cupyx.lapack.gesv` in `cupy.linalg.inv` and reduce code duplication.

toslunar reviewed (approved):
LGTM.  BTW, it should be discussed later whether the current specification of `cupyx.lapack.gesv` (`promote_types` and copy are there) is good.

toslunar commented:
pfnCI, test this please.

anaruse commented:
> BTW, it should be discussed later whether the current specification of cupyx.lapack.gesv (promote_types and copy are there) is good.

I think what you mean is that, it is better 
* it is better not to promote data type in `cupyx.lapack.gesv`, and
* it is OK to overwrite arrays rather than copying them to prevent them to be overwritten in `cupyx.lapack.gesv`,
is this correct?

If so, I would like to make the correction in this PR, is that OK? That's because the failures currently observed in this PR are related to this.

anaruse commented:
I changed `cupyx.lapack.gesv` so that it does not promote dtype or copy the array (as much as possible). I also changed `cupy.linalg.solve` and `cupy.linalg.inv` accordingly.
Could you take a look again?

toslunar commented:
I'm sorry but let me suspend the review until #4309 is fixed for v8.

anaruse commented:
> I'm sorry but let me suspend the review until #4309 is fixed for v8.

Got it, but probably the issue you saw in #4309 has been solved in this PR.

toslunar commented on cupyx/lapack.py:
```diff
@@ -39,21 +43,22 @@ def gesv(a, b):
+        if not a._f_contiguous:
+            a = a.copy(order='F')
+    ret_b = b
+    if nrhs > 1 and not ret_b._f_contiguous:
```
`b` can be non-contiguous even if `nrhs <= 1`.
```suggestion
    if not ret_b._f_contiguous:
```

toslunar commented on cupyx/lapack.py:
```diff
@@ -66,11 +71,13 @@ def gesv(a, b):
     _cupy.linalg._util._check_cusolver_dev_info_if_synchronization_allowed(
         getrs, dinfo)
-    return b
+    if nrhs > 1 and not ret_b._f_contiguous:
```
ditto

anaruse replied:
And no need to return `b`?

toslunar reviewed (commented):
In `cupyx.lapack.gesv`, let's avoid copy by raising `ValueError` unless
- `a` is C- or F-contiguous and
- `b` is F-contiguous

toslunar commented:
> And no need to return `b`?

Exactly!

anaruse commented:
Thanks for your comment. I've fixed it based on your feedback. Could you take another look?

toslunar commented on cupyx/lapack.py:
```diff
@@ -50,15 +50,15 @@ def gesv(a, b):
+        raise ValueError('a must be F-contiguous or C-contiguous.')
+    if not (b._f_contiguous or (b._c_contiguous and nrhs == 1)):
+        raise ValueError('b must be F-contiguous '
+                         '(or C-contiguous if nrhs == 1).')
```
The message looks good.

For the `nrhs == 1` cases, I suppose `b._f_contiguous` iff `b._c_contiguous`.

anaruse replied:
Thanks, I did not know that.
Should I change (simplify) the `if` condition?

toslunar replied:
Shorter error message looks better here. Applying `asfortranarray` to a C- and F-contiguous array is redundant, but has no GPU cost.

toslunar commented on cupyx/lapack.py:
```diff
@@ -39,21 +43,22 @@ def gesv(a, b):
+        trans = _cublas.CUBLAS_OP_T
+    else:
+        raise ValueError('a must be F-contiguous or C-contiguous.')
+    if not (b._f_contiguous or (b._c_contiguous and nrhs == 1)):
+        raise ValueError('b must be F-contiguous '
+                         '(or C-contiguous if nrhs == 1).')
```
```suggestion
    if not b._f_contiguous:
        raise ValueError('b must be F-contiguous.')
```

toslunar commented on tests/cupyx_tests/test_lapack.py:
```diff
@@ -30,36 +32,59 @@ def _make_matrix(self, shape, alpha, beta):
+    def test_invalid_cases(self):
+        if self.nrhs is None or self.nrhs == 1:
+            raise unittest.SkipTest()
+        with pytest.raises(ValueError):
+            ng_a = self.a.reshape(1, self.n, self.n)
```
Could you create `ng_a` (and `ng_b`) outside `with pytest.raises`?

anaruse replied:
OK, I will move them outside `with pytest.raise`.

toslunar commented:
pfnCI, test this please.

anaruse commented:
Jenkins, test this please.

toslunar reviewed (approved):
LGTM

toslunar commented:
/test

kmaehashi commented:
@anaruse You can make your [CuPy organization visibility](https://github.com/orgs/cupy/people) from private to public to trigger `pfn-public-ci`s.

toslunar commented:
Backport of #4363 might cherry-pick some commits here.
